lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,239 @@
+					// handling global requests.
+					log.WarnContext(ctx, "Failed to forward global SSH request to target", "error", err)
+					continue
+				}
+				err = trace.Wrap(err, "forwarding global request to target")
+				if replyErr := req.Reply(false, nil); replyErr != nil {
+					// A reply was expected and we failed to send one, we're in a
+					// bad state, return an error to terminate the connection.
+					return trace.NewAggregate(err, trace.Wrap(replyErr, "replying to request with error"))
+				}
+				// We failed to send the request, but we informed the client of
+				// the failure with req.Reply(false, nil), so it's safe to
+				// continue.
+				continue
+			}
+			if err := req.Reply(ok, reply); err != nil {
+				// A reply was expected by the client and returned by the target
+				// but we failed to forward it back. We're in a bad state,
+				// return an error to terminate the connection.
+				return trace.Wrap(err, "forwarding reply from target back to client")
+			}
+		}
+	}
+}
+
+func proxyChannels(
+	ctx context.Context,
+	targetClient *ssh.Client,
+	channels <-chan ssh.NewChannel,
+) (err error) {
+	// Every channel will be handled in its own goroutine, but a recoverable
+	// error from a single channel should not terminate other channels, so
+	// handleNewChannel must avoid returning errors that should not terminate
+	// the connection.
+	g, ctx := errgroup.WithContext(ctx)
+	defer func() {
+		chanErr := g.Wait()
+		if chanErr == nil ||
+			utils.IsOKNetworkError(chanErr) ||
+			errors.Is(chanErr, context.Canceled) {
+			return
+		}
+		err = trace.NewAggregate(err, chanErr)
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case newChan, ok := <-channels:
+			if !ok {
+				return nil
+			}
+			g.Go(func() error {
+				return trace.Wrap(handleNewChannel(ctx, targetClient, newChan),
+					"unexpected error while handling SSH channel")
+			})
+		}
+	}
+}
+
+func handleNewChannel(ctx context.Context, targetClient *ssh.Client, newChan ssh.NewChannel) error {
+	log.DebugContext(ctx, "Handling new SSH channel", "type", newChan.ChannelType())
+
+	// Try to open a corresponding channel on the target. If the target rejects
+	// the channel, reject the incoming channel request.
+	targetChan, targetChanRequests, err := targetClient.OpenChannel(newChan.ChannelType(), newChan.ExtraData())
+	if err != nil {
+		reason, message, unexpectedOpenChannelErr := rejectChannelReason(err)
+		rejectErr := trace.Wrap(newChan.Reject(reason, message),
+			"rejecting incoming channel request after failing to open channel on target")
+		if unexpectedOpenChannelErr != nil || rejectErr != nil {
+			// Either of these errors is unexpected so return an error
+			// to terminate the full SSH connection.
+			return trace.NewAggregate(unexpectedOpenChannelErr, rejectErr)
+		}
+		// The channel was successfully rejected, this is fine, the client may
+		// have requested a channel type the target doesn't support.
+		return nil
```

Comment:
This whole error handling logic is a bit complex to follow, I'd like to see the `OpenChannelError` case (which at the "forwarder" level like this is not actually an error condition at all) more distinctly split from the hard error case.

Reply:
got rid of the weird helper function that didn't actually help